web/modules/custom/joinup_discussion/src/EventSubscriber/SubscribedDiscussionSubscriber.php

```diff
@@ -76,56 +99,159 @@ public static function getSubscribedEvents(): array {
-      '@group:label' => $this->group->label(),
-      '@group:bundle' => $this->group->bundle(),
-      '@entity:url' => $this->discussion->toUrl('canonical', [
-        'absolute' => TRUE,
-      ])->toString(),
-    ];
+  protected function getArguments(NodeInterface $discussion) : array {
+    $arguments = [];
+
+    $arguments['@entity:title'] = $discussion->label();
+    try {
+      $entity_url = $discussion->toUrl('canonical', ['absolute' => TRUE])->toString();
+    }
+    catch (EntityMalformedException | UndefinedLinkTemplateException $e) {
+      // Retrieval of the URL might fail if the link template is not defined, or
+      // if the entity doesn't have an ID.
+      throw new \Exception('Could not retrieve URL for discussion.', NULL, $e);
+    }
+    $arguments['@entity:url'] = $entity_url;
+
+    $actor = $this->getCurrentUser();
+    $actor_first_name = !empty($actor->get('field_user_first_name')->value) ? $actor->get('field_user_first_name')->value : '';
+    $actor_family_name = !empty($actor->get('field_user_family_name')->value) ? $actor->get('field_user_family_name')->value : '';
+
+    if ($actor->hasRole('moderator')) {
+      $arguments['@actor:full_name'] = 'The Joinup Support Team';
+    }
+    else {
+      $arguments['@actor:full_name'] = empty($actor->get('full_name')->value) ? $actor_first_name . ' ' . $actor_family_name : $actor->get('full_name')->value;
+    }
+
+    $group = $this->getDiscussionGroup($discussion);
+    if ($group) {
+      $arguments['@group:label'] = $group->label();
+      $arguments['@group:bundle'] = $group->bundle();
+    }
+
+    return $arguments;
+  }
+
+  /**
+   * Returns the fully loaded User entity for the current user.
+   *
+   * @return \Drupal\user\UserInterface
+   *   The current user.
+   */
+  protected function getCurrentUser(): UserInterface {
+    try {
+      /** @var \Drupal\user\UserInterface $user */
+      $user = $this->entityTypeManager->getStorage('user')->load($this->currentUser->id());
+      return $user;
+    }
+    catch (InvalidPluginDefinitionException $e) {
+      // The storage for the 'user' entity type should exist. Ignore this.
+    }
+  }
+
+  /**
+   * Sends the notification to the recipients.
+   *
+   * @param \Drupal\node\NodeInterface $discussion
+   *   The discussion for which to send the notification.
+   * @param string $message_template
+   *   The ID of the message template to use.
+   *
+   * @return bool
+   *   Whether or not the sending of the e-mails has succeeded.
+   */
+  protected function sendMessage(NodeInterface $discussion, string $message_template): bool {
+    try {
+      return $this->messageDelivery
+        ->createMessage($message_template)
+        ->setArguments($this->getArguments($discussion))
+        ->setRecipients($this->getSubscribers($discussion))
+        ->sendMail();
+    }
+    catch (\Exception $e) {
+      $this->logger->critical('Unexpected exception thrown when sending a message for a discussion.',
```

Comment:
Why using the `'php'` channel here? Wouldn't be more appropriate to use the `'mail'` channel as `MailManager` is doing. And I like that they are injecting the logger factory instead a specific logger channel.

Reply:
https://github.com/drupal/core/blob/8.5.x/lib/Drupal/Core/Mail/MailManager.php#L300

Reply:
Very good point, I wasn't aware of the existence of this channel. I copied this approach from somewhere. I'll also have a look at injecting the factory, that sounds like a clean solution.